### ISSUE TITLE
race conditions in the code.

### ISSUE THREAD

**Issue by vondele** [nodchip/Stockfish#164]
current gensfen and learn commands are not correct multithreaded, several races remain. To check:

```
make clean && make -j2 ARCH=x86-64-modern sanitize=thread    optimize=no debug=yes build > /dev/null && ../tests/instrumented_learn.sh --sanitizer-thread
```

**Comment by vondele** [comment]
@Sopel97 maybe you can test if this is fixed by adding that line to .travis.yml ?

**Comment by Sopel97** [comment]
I've had some weird stuff happen with a certain dataset. I have a dataset that has qsearch PV almost always be empty. With this dataset I'm observing randomly (~50% of the time) training getting completely borked at the start. Sometimes it crashes during calc_loss and sometimes it just outputs norm being 2000*startpos_eval. I have no idea what can cause that. I'm gonna leave the note here because I strongly suspect it's some kind of a race that happens before/during the first call to calc_loss.

**Comment by Sopel97** [comment]
@nodchip 
I believe the above issue is from the fact that `Position::set_state` doesn't do `st->accumulator.computed_accumulation = false;`. It's only done in `do_move`. If in training qsearch returns no PV then the following Eval::NNUE:evaluate() will return previous evaluation which will be wrong

**Comment by vondele** [comment]
@nodchip @Sopel this was auto-closed on merge, maybe we can explicitly add the line above to CI, to make sure it is fixed.

**Comment by Sopel97** [comment]
Right, not sure why this was closed straight up. There may still be issues in gensfen. I'd reopen this for now, later I'll try adding the CI tasks when I reduce the bench depth in valgrind.

**Comment by vondele** [comment]
doesn't look like I can reopen @nodchip should have the permission to do that.

git will auto-close issues if the commit message says `fixes #1234` unfortunately also for 'maybe fixes' 'partially fixes' etc.
